Add doc to explain how config container-freezer autoscaling (#4633)
knative/docs: docs/serving/autoscaling/container-freezer.md

# Configuring container-freezer

> When container-freezer is enabled, queue-proxy will call an endpoint API when its traffic drops to zero or scales up from zero.

Author: fixed

> Within official endpoint API implementation container-freezer, the running process will be frozen when the pod's traffic drop to zero and resumed when the pod's traffic scales up from zero. The endpoint API is user-specific implementation, we can even implement it as a billing component, because it knows when the requests are being handled.

Author: It doesn't matter to do it before or after, both works.

## Configure min-scale

> If we want to use container-freezer, the value of per-revision annotation key `autoscaling.knative.dev/min-scale` should be bigger than 0.

Author: fixed

**Example:**
=== "Per Revision"
    ```yaml
    apiVersion: serving.knative.dev/v1
    kind: Service
    metadata:
      name: helloworld-go
      namespace: default
    spec:
      template:
        metadata:
          annotations:
            autoscaling.knative.dev/min-scale: "3"
        spec:
          containers:
            - image: gcr.io/knative-samples/helloworld-go
    ```

> ## Configure endpoint api address

Author: fixed

> queue-proxy will call endpoint api address when container-freezer is enabled, so we need to configure the api address if we want to use container-freezer.

Author: fixed

> The endpoint api address can be configured by:
> ```bash
> kubectl edit configmap config-deployment -n knative-serving
> ```
> 
> queue-proxy will swap it as the node IP if using `$HOST_IP` in configuration, because the official implementation is a daemonset. If we implement the endpoint api as a service in the cluster, there should be a specific service address, such as `http://billing.default.svc:9696`.
> **Example:**
> ```yaml
> apiVersion: v1
> kind: ConfigMap
> metadata:
>   name: config-deployment
>   namespace: knative-serving
> data:
>   concurrency-state-endpoint: "http://$HOST_IP:9696"
> ```

Reviewer: ```suggestion
1. Open the `config-deployment` ConfigMap by running the command:

    ```bash
    kubectl edit configmap config-deployment -n knative-serving
    ```

1. Edit the file to configure the endpoint API address, for example:

    ```yaml
    apiVersion: v1
    kind: ConfigMap
    metadata:
      name: config-deployment
      namespace: knative-serving
    data:
      concurrency-state-endpoint: "http://$HOST_IP:9696"
    ```
    !!! note
        If you use `$HOST_IP`, queue-proxy inserts the appropriate value for each node because the official
implementation is a daemonset. If you implement the endpoint API as a service in the cluster, use a specific service address such as `http://billing.default.svc:9696`.
```

Reviewer: This would be better as steps (as suggested above).

Author: fixed

## Next
* Implement your own user-specific endpoint api and deploying it in cluster.

> * Using the official implementation [container-freezer](https://github.com/knative-sandbox/container-freezer).

Author: fixed

# Configuring container-freezer

When container-freezer is enabled, queue-proxy calls an endpoint API when its traffic drops to zero or scales up from zero.

> Within the official endpoint API implementation container-freezer, the running process is frozen when the pod's traffic drops to zero, and resumed when the pod's traffic scales up from zero. The endpoint API is a user-specific implementation. For example, you can implement it as a billing component because it knows when the requests are being handled.

Reviewer 2: ```suggestion
Within the official endpoint API implementation container-freezer, the running process is frozen when the pod's traffic drops to zero, and resumed when the pod's traffic scales up from zero. However, users are also run their own  implementation instead (for example, as a billing component to log when requests are being handled).
```

Author: fixed

> Within the official endpoint API implementation container-freezer, the running process is frozen when the pod's traffic drops to zero, and resumed when the pod's traffic scales up from zero. The endpoint API is a user-specific implementation. For example, you can implement it as a billing component because it knows when the requests are being handled.

Reviewer 2: Also, and this is a nit, I wonder if "official" is the right word here... maybe "community-maintained" or something similar?

Author: community-maintained is a good option

Author: fixed

## Configure min-scale

To use container-freezer, the value of per-revision annotation key `autoscaling.knative.dev/min-scale` must be greater than zero.

**Example:**
=== "Per Revision"
    ```yaml
    apiVersion: serving.knative.dev/v1
    kind: Service
    metadata:
      name: helloworld-go
      namespace: default
    spec:
      template:
        metadata:
          annotations:
            autoscaling.knative.dev/min-scale: "3"
        spec:
          containers:
            - image: gcr.io/knative-samples/helloworld-go
    ```

## Configure the endpoint API address

> queue-proxy calls the endpoint API address when container-freezer is enabled, so you need to configure the API address.

Reviewer 2: ```suggestion
Queue-proxy calls the endpoint API address when container-freezer is enabled, so you need to configure the API address.
```

(it's been a while, but I think capitalizing here is correct)

1. Open the `config-deployment` ConfigMap by running the command:
    ```bash
    kubectl edit configmap config-deployment -n knative-serving
    ```

> 2. Edit the file to configure the endpoint API address, for example:

Reviewer 2: ```suggestion
1. Edit the file to configure the endpoint API address, for example:
```

Author: fixed

    ```yaml
    apiVersion: v1
    kind: ConfigMap
    metadata:
      name: config-deployment
      namespace: knative-serving
    data:
      concurrency-state-endpoint: "http://$HOST_IP:9696"
    ```

    !!! note

>         If you use `$HOST_IP`, queue-proxy inserts the appropriate value for each node because the official implementation is a daemonset. If you implement the endpoint API as a service in the cluster, use a specific service address such as `http://billing.default.svc:9696`.

Reviewer 2: I'd reverse the order for the first sentence, something like:
```markdown
If using the official implementation, use `http://$HOST_IP:9696` as the value for 
`concurrency-state-endpoint`, as the official implementation is a daemonset and
the appropriate value will be inserted by queue-proxy at runtime.
```

Reviewer 3: Let's use something else than `official` 
Maybe `community reference implementation`
